backend/src/v2/metadata/client.go

```diff
@@ -285,6 +288,26 @@ func (e *Execution) FingerPrint() string {
 	return e.execution.GetCustomProperties()[keyCacheFingerPrint].GetStringValue()
 }
 
+// GenerateOutputURI appends the specified paths to the pipeline root.
+// It may be configured to preserve the query part of the pipeline root
+// by splitting it off and appending it back to the full URI.
+func GenerateOutputURI(pipelineRoot string, paths []string, preserveQueryString bool) string {
+	querySplit := strings.Split(pipelineRoot, "?")
+	query := ""
+	if len(querySplit) == 2 {
+		pipelineRoot = querySplit[0]
+		if preserveQueryString {
+			query = "?" + querySplit[1]
+		}
+	} else if len(querySplit) > 2 {
+		// this should never happen, but just in case.
+		glog.Warningf("Unexpected pipeline root: %v", pipelineRoot)
+	}
+	// we cannot path.Join(root, taskName, artifactName), because root
+	// contains scheme like gs:// and path.Join cleans up scheme to gs:/
+	return fmt.Sprintf("%s/%s%s", strings.TrimRight(pipelineRoot, "/"), path.Join(paths...), query)
+}
+
 // GetPipeline returns the current pipeline represented by the specified
 // pipeline name and run ID.
 func (c *Client) GetPipeline(ctx context.Context, pipelineName, runID, namespace, runResource, pipelineRoot, bucketSessionInfo string) (*Pipeline, error) {
```

Comment:
this differs from upstream 2.2 https://github.com/kubeflow/pipelines/blob/2.2.0/backend/src/v2/metadata/client.go#L80

should be storeSessionInfo not bucketSessionInfo

Reply:
some with this: https://github.com/opendatahub-io/data-science-pipelines/pull/78/files#diff-87ec6fcace7656e2c64443ea8b5386b8d54215eab715ab193fc0e2c1bfe939d1R219

this should be GetStoreSessionInfo() https://github.com/kubeflow/pipelines/blob/2.2.0/backend/src/v2/metadata/client.go#L203

Reply:
also this: https://github.com/rimolive/data-science-pipelines/blob/rebase-2.2.0/backend/src/v2/metadata/client.go#L313